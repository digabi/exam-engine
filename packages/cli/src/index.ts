#!/usr/bin/env node

import { accessSync } from 'fs'
import ora = require('ora')
import yargs from 'yargs'
import { resolveExam, resolveFile } from './utils'

/* eslint-disable @typescript-eslint/no-misused-promises */
yargs
  .usage('Usage: $0 <command> [options]')
  .command(
    'new <directory>',
    'Create a new exam',
    (argv) => {
      argv
        .positional('directory', {
          description: 'The directory for the new exam',
          type: 'string',
        })
        .demandOption('directory')
    },
    runCommand('./commands/new')
  )
  .command(
    ['preview [exam] [options]', 'start'],
    'Preview an exam',
    (argv) => {
      addExamArgs(argv)
      argv.option('port', {
        alias: 'p',
        description: 'The port to use in the preview HTTP server',
        type: 'number',
        default: 0,
      })
    },
    runCommand('./commands/preview')
  )
  .command(
    'create-transfer-zip [exam] [options]',
    'Create a transfer zip that can be imported to Oma Abitti',
    addExamArgs,
    runCommand('./commands/create-transfer-zip')
  )
  .command(
    'create-offline [exam] [options]',
    'Create a standalone offline version of the exam.',
    (argv) => {
      addExamAndOutdirArgs(argv)
      argv.option('media', {
        description:
          'Create a media version of the exam. This will encode video files as x264 and audio files as mp3 and not remove hidden references from the exam.',
        type: 'boolean',
      })
    },
    runCommand('./commands/create-offline')
  )
  .command(
    'create-mex [exam] [options]',
    'Package the exam to a .mex file that can be imported by Abitti',
    (argv) => {
      addExamAndOutdirArgs(argv)
      argv
        .option('private-key', {
          alias: 'k',
          description: 'The private key (in PEM format) used for signing the exam',
          coerce: resolveFile,
        })
        .option('passphrase', {
          alias: 'p',
          description: 'The secret passphrase that decrypts the exam',
        })
        .option('nsa-scripts', {
          alias: 'n',
          description: 'Monitoring scripts as a .zip file',
          coerce: resolveFile,
        })
        .option('security-codes', {
          alias: 's',
          description: 'The security codes (in JSON format)',
          coerce: resolveFile,
        })
        .demandOption(['private-key', 'passphrase', 'nsa-scripts'])
    },
    runCommand('./commands/create-mex')
  )
  .check((argv: { exam?: string }) => {
    if (argv.exam) {
      accessSync(argv.exam)
    }
    return true
  })
  .demandCommand()
  .strict()
  .wrap(yargs.terminalWidth()).argv

/* eslint-enable @typescript-eslint/no-misused-promises */

function addExamArgs(yargv: yargs.Argv) {
  yargv.positional('exam', {
    description: 'The path to an exam XML file or the directory containing it.',
    default: process.cwd(),
    coerce: resolveExam,
  })
}

function addExamAndOutdirArgs(argv: yargs.Argv) {
  addExamArgs(argv)
  argv.option('outdir', {
    alias: 'o',
    description: 'The output directory',
    type: 'string',
    coerce: resolveFile,
  })
}

function runCommand<T>(moduleName: string) {
  return async (args: T) => {
    const spinner = ora().start()
    try {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-assignment
      const command = require(moduleName).default
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-call
      const result = await command({ ...args, spinner })
      if (result) {
        spinner.succeed(result)
      }
    } catch (err) {
      spinner.fail((err as Error).stack)
      process.exit(1)
    } finally {
      spinner.stop()
    }
  }
}
