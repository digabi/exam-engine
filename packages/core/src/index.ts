import '../src/css/main.less'
import Attachments from './components/attachments/Attachments'
import Exam from './components/exam/Exam'
import GradingInstructions from './components/grading-instructions/GradingInstructions'
import Results from './components/results/Results'
import parseExam from './parser/parseExam'

export { Attachments, Exam, GradingInstructions, Results, parseExam }
export { ExamComponentProps } from './createRenderChildNodes'
export * from './types/ExamAnswer'
export * from './types/ExamServerAPI'
export * from './types/GradingStructure'
export * from './types/Score'

export interface ExamBundle {
  Attachments: typeof Attachments
  Exam: typeof Exam
  parseExam: typeof parseExam
  Results: typeof Results
}
