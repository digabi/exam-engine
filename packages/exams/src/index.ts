import * as glob from 'glob'
import path from 'path'

/** Returns an array of all exam xml files in the exams directory */
export function listExams(): string[] {
  return glob.globSync(resolveExam('*/*.xml')).sort()
}

/** Resolves a filename relative to the exams directory */
export function resolveExam(filename: string): string {
  return path.resolve(__dirname, '..', filename)
}
