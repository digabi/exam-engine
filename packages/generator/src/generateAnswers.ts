import { ChoiceGroupQuestion, ExamAnswer, GradingStructure, TextQuestion } from '@digabi/exam-engine-core'
import _ from 'lodash'

export function generateAnswers(gradingStructure: GradingStructure): ExamAnswer[] {
  return _.flatMap(gradingStructure.questions, question => {
    switch (question.type) {
      case 'text':
        return generateTextAnswer(question)
      case 'choicegroup':
        return generateChoiceAnswer(question)
    }
  })
}

function generateTextAnswer(question: TextQuestion): ExamAnswer[] {
  const value = `Answer to question ${question.displayNumber}`
  return [
    {
      questionId: question.id,
      type: 'text',
      characterCount: getCharacterCount(value),
      value
    }
  ]
}

function generateChoiceAnswer(question: ChoiceGroupQuestion): ExamAnswer[] {
  return question.choices.map((choice, i) => {
    const option = choice.options[i % choice.options.length]
    return {
      questionId: choice.id,
      type: 'choice',
      value: String(option.id)
    }
  })
}

function getCharacterCount(answerText: string) {
  return answerText.replace(/\s/g, '').length
}
