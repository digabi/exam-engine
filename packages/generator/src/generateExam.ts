import * as libxml from 'libxmljs2'

export interface GenerateExamOptions {
  date?: string
  examCode?: string
  dayCode?: string
  maxAnswers?: number
  languages?: string[]
  sections: GenerateSectionOptions[]
}

export interface GenerateSectionOptions {
  maxAnswers?: number
  casForbidden?: boolean
  questions: GenerateQuestionOptions[][]
}

export type GenerateQuestionOptions =
  | GenerateQuestionOptions[]
  | GenerateTextAnswerOptions
  | GenerateScoredTextAnswerOptions
  | GenerateChoiceAnswerOptions
  | GenerateDropdownAnswerOptions

interface TextAnswerBase {
  hint?: string
  type?: 'rich-text' | 'single-line' | 'multi-line'
}

export interface GenerateTextAnswerOptions extends TextAnswerBase {
  name: 'text-answer'
  maxScore: number
}

export interface GenerateScoredTextAnswerOptions extends TextAnswerBase {
  name: 'scored-text-answer'
  maxScore?: number
  acceptedAnswers?: GenerateAcceptedAnswer[]
}

export interface GenerateAcceptedAnswer {
  score: number
  text: string
}

export interface GenerateChoiceAnswerOptions {
  name: 'choice-answer'
  options: GenerateChoiceAnswerOption[]
}

export interface GenerateDropdownAnswerOptions {
  name: 'dropdown-answer'
  options: GenerateChoiceAnswerOption[]
}

export interface GenerateChoiceAnswerOption {
  score: number
  text: string
}

/**
 * Generates an exam XML file (mostly for testing purposes) based on a
 * description of the exam structure.
 *
 * - An exam is an object containing the array of sections and optional attributes.
 * - A section is object containing the array of questions and optional attributes.
 * - A question is an array of answers or subquestions.
 * - An answer is an object describing the answer.
 *
 * ```
 * generateExam({
 *   sections: [
 *     // Section I
 *     {
 *       questions: [
 *         // Question 1 containing a single text-answer.
 *         [textAnswer()],
 *         // Question 2 containing three choice-answers.
 *         [choiceAnswer(), choiceAnswer(), choiceAnswer()],
 *         // Question 3 containing subquestions
 *         [
 *            // Question 3.1, containing a single dropdown-answer
 *            [dropdownAnswer()],
 *         ]
 *       ]
 *     }
 * ]
 * })
 * ```
 *
 * You can customize the exam and section elements by providing some of the
 * optional attributes.
 *
 * ```
 * generateExam({
 *   examCode: 'M',
 *   date: '2020-03-18',
 *   sections: [{
 *     casForbidden: true,
 *     maxAnswers: 5,
 *     questions: […]
 *   }, {
 *     casForbidden: false,
 *     maxAnswers: 5,
 *     questions: […]
 *   }]
 * })
 * ```
 *
 * It is also possible to customize the answer elements. For example, if you
 * need a text-answer with a different max-score, you can override the default
 * attributes in the various fooAnswer-functions.
 *
 * ```
 * generateExam({ sections: [{ questions: [[textAnswer({ maxScore: 10 })]] }] })
 * ```
 */
export function generateExam(options: GenerateExamOptions): string {
  const doc = new libxml.Document()
  const exam = createElement(doc, 'exam', undefined, {
    xmlns: 'http://www.w3.org/1999/xhtml',
    'xmlns:xsi': 'http://www.w3.org/2001/XMLSchema-instance',
    'xsi:schemaLocation': 'http://ylioppilastutkinto.fi/exam.xsd https://abitti.dev/schema/exam.xsd',
    'exam-schema-version': '0.1',
    'exam-code': options.examCode,
    'day-code': options.dayCode,
    date: options.date,
    'max-answers': options.maxAnswers,
  })
  const languages = createElement(exam, 'languages')
  for (const language of options.languages ?? ['fi-FI']) {
    createElement(languages, 'language', language)
  }
  createElement(exam, 'exam-title', 'Kokeen otsikko')

  for (const section of options.sections) {
    addSection(exam, section)
  }

  return doc.toString(true)
}

function addSection(exam: libxml.Element, options: GenerateSectionOptions): void {
  const section = createElement(exam, 'section', undefined, {
    'max-answers': options.maxAnswers,
    'cas-forbidden': options.casForbidden,
  })

  createElement(section, 'section-title', 'Osan otsikko')

  for (const question of options.questions) {
    addQuestion(section, question)
  }
}

function addQuestion(parent: libxml.Element, options: GenerateQuestionOptions): void {
  if (Array.isArray(options)) {
    const question = createElement(parent, 'question')
    createElement(question, 'question-title', 'Kysymyksen otsikko')
    for (const subquestionOptions of options) {
      addQuestion(question, subquestionOptions)
    }
  } else {
    switch (options.name) {
      case 'text-answer':
      case 'scored-text-answer':
        return addTextAnswer(parent, options)
      case 'choice-answer':
      case 'dropdown-answer':
        return addChoiceAnswer(parent, options)
    }
  }
}

function addTextAnswer(
  question: libxml.Element,
  options: GenerateTextAnswerOptions | GenerateScoredTextAnswerOptions
): void {
  const answer = createElement(question, options.name, undefined, {
    'max-score': options.maxScore,
    type: options.type,
  })

  if (options.hint) {
    createElement(answer, 'hint', options.hint)
  }

  if (options.name === 'scored-text-answer' && options.acceptedAnswers) {
    for (const { score, text } of options.acceptedAnswers) {
      createElement(answer, 'accepted-answer', text, { score })
    }
  }
}

function addChoiceAnswer(
  question: libxml.Element,
  options: GenerateChoiceAnswerOptions | GenerateDropdownAnswerOptions
): void {
  const answer = createElement(question, options.name)
  const optionType = options.name === 'choice-answer' ? 'choice-answer-option' : 'dropdown-answer-option'

  for (const { text, score } of options.options) {
    createElement(answer, optionType, text, { score })
  }
}

export function textAnswer(options?: Partial<GenerateTextAnswerOptions>): GenerateTextAnswerOptions {
  return {
    name: 'text-answer',
    maxScore: 6,
    type: 'rich-text',
    ...options,
  }
}

export function scoredTextAnswer(options?: Partial<GenerateScoredTextAnswerOptions>): GenerateScoredTextAnswerOptions {
  return {
    name: 'scored-text-answer',
    hint: 'Vihje',
    acceptedAnswers: [
      {
        text: 'Oikea vaihtoehto',
        score: 3,
      },
    ],
    ...options,
  }
}

export function choiceAnswer(options?: Partial<GenerateChoiceAnswerOptions>): GenerateChoiceAnswerOptions {
  return {
    name: 'choice-answer',
    options: [
      {
        text: 'Oikea vaihtoehto',
        score: 3,
      },
      {
        text: 'Väärä vaihtoehto',
        score: 0,
      },
      {
        text: 'Väärä vaihtoehto',
        score: 0,
      },
    ],
    ...options,
  }
}

export function dropdownAnswer(options?: Partial<GenerateDropdownAnswerOptions>): GenerateDropdownAnswerOptions {
  return {
    name: 'dropdown-answer',
    options: [
      {
        text: 'Oikea vaihtoehto',
        score: 3,
      },
      {
        text: 'Väärä vaihtoehto',
        score: 0,
      },
      {
        text: 'Väärä vaihtoehto',
        score: 0,
      },
    ],
    ...options,
  }
}

function createElement(
  parent: libxml.Element | libxml.Document,
  localName: string,
  content?: string,
  attrs?: Record<string, unknown>
): libxml.Element {
  const element = parent.node(localName, content).namespace('e', 'http://ylioppilastutkinto.fi/exam.xsd')

  if (attrs) {
    Object.entries(attrs)
      .filter(([, value]) => value != null)
      .forEach(([name, value]) => element.attr(name, String(value)))
  }

  return element
}
