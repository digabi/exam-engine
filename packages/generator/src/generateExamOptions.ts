import {
  GenerateAnswerOptions,
  GenerateChoiceAnswerOptions,
  GenerateDropdownAnswerOptions,
  GenerateQuestionOptions,
  GenerateScoredTextAnswerOptions,
  GenerateTextAnswerOptions,
} from './generateExam'

/**
 * A shorthand version of the function for creating a question containing the specified answers.
 *
 * @param answers An array of answers.
 * @example question([textAnswer()]) // Equivalent to question({ answers: [textAnswer()] })
 */
export function question(answers: GenerateAnswerOptions[]): GenerateQuestionOptions
/**
 * A shorthand version of the function for creating a question containing the specified child questions.
 *
 * @param questions An array of child questions.
 * @example question([question(…)]) // Equivalent to question({ questions: [question(…)] })
 */
export function question(questions: GenerateQuestionOptions[]): GenerateQuestionOptions
/**
 * Create a question.
 *
 * @param options An object describing the structure of the question.
 * @example question({ maxAnswers: 1, answers: [textAnswer(), textAnswer()] })
 */
export function question(options: GenerateQuestionOptions): GenerateQuestionOptions
export function question(
  args: GenerateQuestionOptions | GenerateAnswerOptions[] | GenerateQuestionOptions[]
): GenerateQuestionOptions {
  if (Array.isArray(args)) {
    if ('answers' in args[0] || 'questions' in args[0]) {
      return { questions: args as GenerateQuestionOptions[] }
    } else {
      return { answers: args as GenerateAnswerOptions[] }
    }
  } else {
    return args
  }
}

export function textAnswer(options?: Partial<GenerateTextAnswerOptions>): GenerateTextAnswerOptions {
  return {
    name: 'text-answer',
    maxScore: 6,
    type: 'rich-text',
    ...options,
  }
}

export function scoredTextAnswer(options?: Partial<GenerateScoredTextAnswerOptions>): GenerateScoredTextAnswerOptions {
  return {
    name: 'scored-text-answer',
    hint: 'Vihje',
    acceptedAnswers: [
      {
        text: 'Oikea vaihtoehto',
        score: 3,
      },
      // The legacy exam generator expects this to be the correct answer. (ノಠ益ಠ)ノ彡┻━┻
      {
        text: 'lol',
        score: 3,
      },
    ],
    ...options,
  }
}

export function choiceAnswer(options?: Partial<GenerateChoiceAnswerOptions>): GenerateChoiceAnswerOptions {
  return {
    name: 'choice-answer',
    options: [
      {
        text: 'Oikea vaihtoehto',
        score: 3,
      },
      {
        text: 'Väärä vaihtoehto',
        score: 0,
      },
      {
        text: 'Väärä vaihtoehto',
        score: 0,
      },
    ],
    ...options,
  }
}

export function dropdownAnswer(options?: Partial<GenerateDropdownAnswerOptions>): GenerateDropdownAnswerOptions {
  return {
    name: 'dropdown-answer',
    options: [
      {
        text: 'Oikea vaihtoehto',
        score: 3,
      },
      {
        text: 'Väärä vaihtoehto',
        score: 0,
      },
      {
        text: 'Väärä vaihtoehto',
        score: 0,
      },
    ],
    ...options,
  }
}
