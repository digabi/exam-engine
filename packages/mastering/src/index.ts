export { createMex, createMultiMex } from './createMex'
export * from './mastering/createGradingStructure'
export { getMediaMetadataFromLocalFile } from './getMediaMetadataFromLocalFile'
export * from './mastering'
export { ns } from './mastering/schema'
export * from './migrations'
