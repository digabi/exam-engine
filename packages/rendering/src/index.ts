export interface RenderingOptions {
  casCountdownDurationSeconds?: number
  openFirefox?: boolean
  port?: number
}

export * from './createOfflineExam'
export * from './previewExam'
