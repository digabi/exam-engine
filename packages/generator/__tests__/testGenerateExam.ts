import {
  choiceAnswer,
  dropdownAnswer,
  GenerateAcceptedAnswer,
  GenerateChoiceAnswerOption,
  GenerateChoiceAnswerOptions,
  GenerateDropdownAnswerOptions,
  generateExam,
  GenerateExamOptions,
  GenerateQuestionOptions,
  GenerateScoredTextAnswerOptions,
  GenerateSectionOptions,
  GenerateTextAnswerOptions,
  question,
  scoredTextAnswer,
  textAnswer,
} from '@digabi/exam-engine-generator'
import { ns, parseExam } from '@digabi/exam-engine-mastering'
import { formatISO } from 'date-fns'
import * as fc from 'fast-check'
import { wrap } from 'jest-snapshot-serializer-raw'
import { Element } from 'libxmljs2'
import { GenerateAnswerOptions, GenerateSubQuestionOptions } from '../src'

describe('generateExam()', () => {
  it('creates a basic exam when called with minimal arguments', () => {
    const exam = generateAndParseExam({ sections: [{ questions: [question([textAnswer()])] }] })
    expect(wrap(exam.toString(false))).toMatchSnapshot()
  })

  it('supports exam-specific attributes', () => {
    const examCode = 'A'
    const dayCode = 'X'
    const date = '2020-03-01'
    const maxAnswers = 5
    const title = 'Otsikko'

    const exam = generateAndParseExam({ examCode, dayCode, date, maxAnswers, title, sections: [{ questions: [] }] })
    const root = exam.root()!

    expect(getAttr(root, 'exam-code')).toEqual(examCode)
    expect(getAttr(root, 'day-code')).toEqual(dayCode)
    expect(getAttr(root, 'date')).toEqual(date)
    expect(getAttr(root, 'max-answers')).toEqual(String(maxAnswers))
    expect(root.get<Element>('//e:exam-title', ns)?.text()).toEqual(title)
  })

  it('creates a section for each entry in the sections array', () => {
    const exam = generateAndParseExam({ sections: [{ questions: [] }, { questions: [] }] })
    expect(exam.find('//e:section', ns)).toHaveLength(2)
  })

  it('supports section attributes', () => {
    const maxAnswers = 5
    const casForbidden = true

    const exam = generateAndParseExam({ sections: [{ maxAnswers, casForbidden, questions: [] }] })
    const section = exam.get<Element>('//e:section', ns)!

    expect(getAttr(section, 'max-answers')).toEqual(String(maxAnswers))
    expect(getAttr(section, 'cas-forbidden')).toEqual(String(casForbidden))
  })

  it('creates a question for each entry in the questions array', () => {
    const exam = generateAndParseExam({
      sections: [{ questions: [question([textAnswer()]), question([textAnswer()])] }],
    })

    expect(exam.find('//e:section/e:question', ns)).toHaveLength(2)
    expect(exam.find('//e:section/e:question/e:text-answer', ns)).toHaveLength(2)
  })

  it('supports subquestions', () => {
    const exam = generateAndParseExam({
      sections: [{ questions: [question([question([question([textAnswer()])])])] }],
    })

    expect(exam.find('//e:question/e:question[1]/e:question[1]/e:text-answer', ns)).toHaveLength(1)
  })

  it('supports creating a text-answer with the default attributes', () => {
    const exam = generateAndParseExam({ sections: [{ questions: [question([textAnswer()])] }] })
    const answer = exam.get<Element>('//e:text-answer', ns)!

    expect(getAttr(answer, 'max-score')).toEqual('6')
  })

  it('supports text-answer attributes', () => {
    const maxScore = 10
    const hint = 'Foo'

    const exam = generateAndParseExam({ sections: [{ questions: [question([textAnswer({ maxScore, hint })])] }] })
    const answer = exam.get<Element>('//e:section/e:question/e:text-answer', ns)!

    expect(getAttr(answer, 'max-score')).toEqual(String(maxScore))
    expect(answer.get<Element>('.//e:hint', ns)?.text()).toEqual(hint)
  })

  it('supports creating a scored-text-answer with the default attributes', () => {
    const exam = generateAndParseExam({ sections: [{ questions: [question([scoredTextAnswer()])] }] })
    const acceptedAnswer = exam.get<Element>('//e:section/e:question/e:scored-text-answer/e:accepted-answer', ns)!

    expect(acceptedAnswer.text()).toEqual('Oikea vaihtoehto')
    expect(getAttr(acceptedAnswer, 'score')).toEqual('3')
  })

  it('supports scored-text-answer attributes', () => {
    const maxScore = 10
    const hint = 'Foo'
    const acceptedAnswers = [
      { text: 'Bar', score: 1 },
      { text: 'Baz', score: 2 },
    ]

    const exam = generateAndParseExam({
      sections: [{ questions: [question([scoredTextAnswer({ maxScore, hint, acceptedAnswers })])] }],
    })

    const answer = exam.get<Element>('//e:section/e:question/e:scored-text-answer', ns)!

    expect(getAttr(answer, 'max-score')).toEqual(String(maxScore))
    expect(answer.get<Element>('.//e:hint/text()', ns)?.toString()).toEqual(hint)
    expect(
      answer
        .find<Element>('.//e:accepted-answer', ns)
        .map((e) => ({ text: e.text(), score: Number(getAttr(e, 'score')) }))
    ).toEqual(acceptedAnswers)
  })

  it('supports creating a choice-answer with the default attributes', () => {
    const exam = generateAndParseExam({ sections: [{ questions: [question([choiceAnswer()])] }] })
    const options = exam
      .find<Element>('//e:section/e:question/e:choice-answer/e:choice-answer-option', ns)
      .map((e) => ({ text: e.text(), score: Number(getAttr(e, 'score')) }))

    expect(options).toEqual([
      { text: 'Oikea vaihtoehto', score: 3 },
      { text: 'Väärä vaihtoehto', score: 0 },
      { text: 'Väärä vaihtoehto', score: 0 },
    ])
  })

  it('supports choice-answer attributes', () => {
    const expected = [
      { text: 'Bar', score: 1 },
      { text: 'Baz', score: 2 },
    ]

    const exam = generateAndParseExam({ sections: [{ questions: [question([choiceAnswer({ options: expected })])] }] })
    const options = exam
      .find<Element>('//e:section/e:question/e:choice-answer/e:choice-answer-option', ns)
      .map((e) => ({ text: e.text(), score: Number(getAttr(e, 'score')) }))

    expect(options).toEqual(expected)
  })

  it('supports creating a dropdown-answer with the default attributes', () => {
    const exam = generateAndParseExam({ sections: [{ questions: [question([dropdownAnswer()])] }] })
    const options = exam
      .find<Element>('//e:section/e:question/e:dropdown-answer/e:dropdown-answer-option', ns)
      .map((e) => ({ text: e.text(), score: Number(getAttr(e, 'score')) }))

    expect(options).toEqual([
      { text: 'Oikea vaihtoehto', score: 3 },
      { text: 'Väärä vaihtoehto', score: 0 },
      { text: 'Väärä vaihtoehto', score: 0 },
    ])
  })

  it('supports dropdown-answer attributes', () => {
    const expected = [
      { text: 'Foo', score: 1 },
      { text: 'Bar', score: 2 },
    ]

    const exam = generateAndParseExam({
      sections: [{ questions: [question([dropdownAnswer({ options: expected })])] }],
    })
    const options = exam
      .find<Element>('//e:section/e:question/e:dropdown-answer/e:dropdown-answer-option', ns)
      .map((e) => ({ text: e.text(), score: Number(getAttr(e, 'score')) }))

    expect(options).toEqual(expected)
  })

  it('creates valid exams', () => {
    const maxScore = fc.integer(1, 10)
    const score = fc.integer(-10, 10)
    const positiveScore = fc.integer(1, 10)
    const maxAnswers = fc.integer(1, 10)
    const type = fc.constantFrom('rich-text' as const, 'single-line' as const)

    const hint = fc.string()
    const textAnswer: fc.Arbitrary<GenerateTextAnswerOptions> = fc.record({
      name: fc.constant('text-answer' as const),
      maxScore,
      hint: optional(hint),
      type: optional(type),
    })
    const acceptedAnswers: fc.Arbitrary<GenerateAcceptedAnswer[]> = fc.array(
      fc.record({ score: positiveScore, text: fc.string() }),
      1,
      3
    )
    const scoredTextAnswer: fc.Arbitrary<GenerateScoredTextAnswerOptions> = fc.oneof(
      fc.record({
        name: fc.constant('scored-text-answer' as const),
        maxScore,
        hint: optional(hint),
        type: optional(type),
      }),
      fc.record({
        name: fc.constant('scored-text-answer' as const),
        hint: optional(hint),
        type: optional(type),
        acceptedAnswers,
      })
    )
    const options: fc.Arbitrary<GenerateChoiceAnswerOption[]> = fc.array(
      fc.record({
        score,
        text: fc.string(),
      }),
      1,
      10
    )
    const choiceAnswer: fc.Arbitrary<GenerateChoiceAnswerOptions> = fc.record({
      name: fc.constant('choice-answer' as const),
      options,
    })
    const dropdownAnswer: fc.Arbitrary<GenerateDropdownAnswerOptions> = fc.record({
      name: fc.constant('dropdown-answer' as const),
      options,
    })
    const answer: fc.Arbitrary<GenerateAnswerOptions> = fc.oneof(
      textAnswer,
      scoredTextAnswer,
      choiceAnswer,
      dropdownAnswer
    )
    const subQuestion: fc.Arbitrary<GenerateSubQuestionOptions> = fc.record({
      maxAnswers: optional(maxAnswers),
      answers: fc.array(answer, 1, 10),
    })
    // Generate Subquestions with 75% probability.
    const question: fc.Memo<GenerateQuestionOptions> = fc.memo<GenerateQuestionOptions>((n) =>
      fc.oneof(rootQuestion(n), subQuestion, subQuestion, subQuestion)
    )
    const rootQuestion: fc.Memo<GenerateQuestionOptions> = fc.memo<GenerateQuestionOptions>((n) => {
      return n === 0
        ? subQuestion
        : fc.record({
            maxAnswers: optional(maxAnswers),
            questions: fc.array(question(n - 1), 1, 5),
          })
    })
    const questions = fc.array(question(2), 1, 5)
    const section: fc.Arbitrary<GenerateSectionOptions> = fc.record({
      maxAnswers: optional(maxAnswers),
      casForbidden: optional(fc.boolean()),
      questions,
    })
    const sections = fc.array(section, 1, 3)
    const exam: fc.Arbitrary<GenerateExamOptions> = fc.record({
      date: optional(fc.date().map((d) => formatISO(d, { representation: 'date' }))),
      examCode: optional(fc.constantFrom('EA', 'M', 'N')),
      dayCode: optional(fc.constant('X')),
      maxAnswers: optional(maxAnswers),
      languages: optional(fc.set(fc.constantFrom('fi-FI', 'sv-FI'), 1, 2)),
      title: optional(fc.string()),
      sections,
    })

    fc.assert(
      fc.property(exam, (examOptions) => {
        expect(() => generateAndParseExam(examOptions)).not.toThrow()
      })
    )
  })
})

function generateAndParseExam(options: GenerateExamOptions) {
  const exam = generateExam(options)
  return parseExam(exam, true)
}

function getAttr(element: Element, name: string) {
  return element.attr(name)?.value()
}

function optional<T>(arb: fc.Arbitrary<T>): fc.Arbitrary<T | undefined> {
  return fc.option(arb, { nil: undefined })
}
