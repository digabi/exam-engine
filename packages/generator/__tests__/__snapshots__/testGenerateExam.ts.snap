// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateExam() creates a basic exam when called with minimal arguments 1`] = `
<?xml version="1.0" encoding="UTF-8"?>
<e:exam xmlns:e="http://ylioppilastutkinto.fi/exam.xsd" xmlns="http://www.w3.org/1999/xhtml" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://ylioppilastutkinto.fi/exam.xsd https://abitti.dev/schema/exam.xsd" exam-schema-version="0.1">
  <e:languages>
    <e:language>fi-FI</e:language>
  </e:languages>
  <e:exam-title>Kokeen otsikko</e:exam-title>
  <e:section>
    <e:section-title>Osan otsikko</e:section-title>
    <e:question>
      <e:question-title>Kysymyksen otsikko</e:question-title>
      <e:text-answer max-score="6" type="rich-text"/>
    </e:question>
  </e:section>
</e:exam>

`;
